### PR TITLE
feat: defiedge token fetchers

### PR DESCRIPTION
**Pull request by Pranavraut033**
## Description

Integrating [defiedge.io](https://defiedge.io/) token fetcher on Ethereum, Polygon, Optimism and Arbitrum.

## Checklist

- [x] I have followed the [Contributing Guidelines](https://github.com/Zapper-fi/studio/blob/main/CONTRIBUTING.md)
- [ ] (optional) As a contributor, my Ethereum address/ENS is:
- [ ] (optional) As a contributor, my Twitter handle is: 

## How to test?

1. Create `.env` with `ENABLED_APPS=defiedge`
2. Run `pnpm dev` to start nest server
3. Visit dev endpoints below to get list of integrate group (strategy):
    - [polygon](http://localhost:5001/apps/defiedge/tokens?groupIds[]=strategy&network=polygon)
    - [optimism](http://localhost:5001/apps/defiedge/tokens?groupIds[]=strategy&network=optimism)
    - [arbitrum](http://localhost:5001/apps/defiedge/tokens?groupIds[]=strategy&network=arbitrum)
    - [ethereum](http://localhost:5001/apps/defiedge/tokens?groupIds[]=strategy&network=ethereum)

### CODE REVIEW

**Review by wpoulin — CHANGES_REQUESTED** [review]

**Comment by wpoulin** [comment]
Here are some comments you will need to address. It will improve the readability of the code quite a lot.
I've also noticed you've added a few more contract's abi than necessary ? I'm wondering if it's because you want to add more positions in the future which will require those contracts ?

**Comment by Pranavraut033** [comment]
Thanks @wpoulin for reviewing the PR. We will make the changes accordingly.

File: src/apps/defiedge/types/defiedge.types.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
+export interface Strategy {
```
[the remark this answers is not in the document]
Comment (Pranavraut033) [reply src/apps/defiedge/types/defiedge.types.ts:1]: removed unused types

File: src/apps/defiedge/common/defiedge.strategy.token-fetcher.ts [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    @Inject(APP_TOOLKIT) private readonly appToolkit: IAppToolkit,
+    @Inject(DefiedgeContractFactory) private readonly defiedgeContractFactory: DefiedgeContractFactory,
+  ) {}
+
+  async getPositions() {
+    const networkParam = this.network === Network.ETHEREUM_MAINNET ? 'mainnet' : this.network;
+    const endpoint = `${DEFIEDGE_BASE_URL}/${networkParam}/strategies`;
+    const multicall = this.appToolkit.getMulticall(this.network);
+
+    const strategies = await axios.get<Strategy[]>(endpoint).then(({ data }) => data);
```
[the remark this answers is not in the document]
Comment (Pranavraut033) [reply src/apps/defiedge/common/defiedge.strategy.token-fetcher.ts]: Code extracted to a function in utils

File: src/apps/defiedge/common/defiedge.strategy.token-fetcher.ts [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+                const baseToken = baseTokenAddressToDetails[t.id.toLowerCase()];
+                const price = baseToken?.price ? baseToken.price : await getTokenPrice(t.symbol);
+
+                return {
+                  address: t.id,
+                  decimals: +t.decimals,
+                  network: this.network,
+                  type: ContractType.BASE_TOKEN,
+                  price: price ?? 0,
+                  symbol: t.symbol,
```
[the remark this answers is not in the document]
Comment (Pranavraut033) [reply src/apps/defiedge/common/defiedge.strategy.token-fetcher.ts]: I have added getBaseToken function to create a BaseToken object if the token is not available in the getBaseTokenPrices array

**Comment by Pranavraut033** [comment]
Hello @wpoulin, I have updated the branch with mentioned changes, please review it and let me know if there are any changes.

Thanks

File: src/apps/defiedge/common/defiedge.strategy.token-fetcher.ts [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        });
+
+        const [decimals, totalSupplyBN, aumWithFee] = await Promise.all([
+          multicall.wrap(erc20Contract).decimals(),
+          multicall.wrap(erc20Contract).totalSupply(),
+          multicall.wrap(strategyContract).callStatic.getAUMWithFees(false),
+        ]);
+
+        const token0 = baseTokenAddressToDetails[strategy.token0.id.toLowerCase()];
+        const token1 = baseTokenAddressToDetails[strategy.token1.id.toLowerCase()];
```
Comment (wpoulin) [inline src/apps/defiedge/common/defiedge.strategy.token-fetcher.ts:61]: Here's what you can do
```
const baseTokens = await this.appToolkit.getBaseTokenPrices(network);
const token0 = baseTokens.find(x => x.address === strategy.token0.id.toLowerCase())
const token1 = baseTokens.find(x => x.address === strategy.token1.id.toLowerCase())
```

You don't need to change the shape of `baseTokens`

File: src/apps/defiedge/common/defiedge.strategy.token-fetcher.ts [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    const multicall = this.appToolkit.getMulticall(this.network);
+
+    const [strategies, baseTokenAddressToDetails] = await Promise.all([
+      getDefiStrategies(this.network),
+      this.appToolkit.getBaseTokenPrices(this.network).then(baseTokens => {
+        return _.keyBy(baseTokens, e => e.address.toLowerCase());
+      }),
+    ]);
+
+    const appTokens = await Promise.allSettled(
```
Comment (wpoulin) [inline src/apps/defiedge/common/defiedge.strategy.token-fetcher.ts:46]: Promise.allSettled isn't necessary, you can replace with `.all`

File: src/apps/defiedge/common/defiedge.strategy.token-fetcher.ts [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        const appToken: AppTokenPosition<StrategyTokenDataProps> = {
+          type: ContractType.APP_TOKEN,
+          address: strategy.id,
+          appId: this.appId,
+          tokens,
+          decimals,
+          groupId: this.groupId,
+          network: this.network,
+          supply,
+          symbol: strategy.title,
```
Comment (wpoulin) [inline src/apps/defiedge/common/defiedge.strategy.token-fetcher.ts:92]: I suggest you get the symbol from the contract and do no overwrite it

File: src/apps/defiedge/common/defiedge.strategy.token-fetcher.ts [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+                value: buildPercentageDisplayItem(sharePrice - 100),
+              },
+            ],
+          },
+        };
+        return appToken;
+      }),
+    );
+
+    return appTokens.map(e => (e.status == 'fulfilled' ? e.value : null)).filter(filterNulls);
```
Comment (wpoulin) [inline src/apps/defiedge/common/defiedge.strategy.token-fetcher.ts:116]: Using `Promise.all`, you will be able to do _.compact(appTokens)

File: src/apps/defiedge/utils/index.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+export function expandTo18Decimals(value: BigNumber, decimals?: number): BigNumber {
+  return value.mul(BigNumber.from(10).pow(BigNumber.from(18).sub(BigNumber.from(decimals ?? 0))));
+}
+
+export function filterNulls<T>(a: T | null): a is T {
+  return Boolean(a);
+}
+
+export async function getDefiStrategies(network: Network) {
```
Comment (wpoulin) [inline src/apps/defiedge/utils/index.ts:18]: In order to increase scallability and performance, you could `@Cache` the response. 
Here's an example : `UnipilotVaultDefinitionsResolver`

**Review by wpoulin — APPROVED** [review]

**Pull request merged**